File: scripts/ns046_dlambda3_sign_check.py
#!/usr/bin/env python3
# ns046_dlambda3_sign_check.py — the NS-046 sign REQUIRED-CHECK, closed exactly (sympy).
#
# WHAT THIS PINS. The NS-046 probes (ns046_{gradxi_pressure,uniform_domination,
# integral_cancellation}) ADOPT the convention "e₃ᵀ(∇²p)e₃ > 0 ⇒ the nonlocal pressure
# DEPLETES the max-stretch growth", resting on the claim Dλ₃/Dt ⊃ −e₃ᵀ(∇²p)e₃ — which was
# never rigorously derived in-repo (flagged as a Required Check in the integral probe).
# This script machine-verifies the full derivation chain SYMBOLICALLY (sympy, exact):
#
#   I1  gradient-of-NS:        D A/Dt = −A² − ∇²p + νΔA          (A_ik = ∂u_i/∂x_k)
#   I2  symmetrization:        sym(A²) = S² + Ω²                  (⇒ DS/Dt = −S² − Ω² − ∇²p + νΔS)
#   I3  rotation term:         Ω² = ¼(ω⊗ω − |ω|² I)              (Ω_ij = −½ ε_ijk ω_k, ω=∇×u)
#   I4  eigen-derivative:      dλ₃/dt = e₃ᵀ (dS/dt) e₃            (simple λ₃; Hellmann–Feynman)
#
#   ⇒  Dλ₃/Dt = −λ₃² + ¼(|ω|² − (ω·e₃)²) − e₃ᵀ(∇²p)e₃ + ν e₃ᵀ(ΔS)e₃
#
# so the pressure-Hessian projection enters with coefficient −1: the adopted convention is
# CORRECT iff every identity below verifies to 0. Caveat carried with the result: I4 requires
# λ₃ SIMPLE (a.e. true; eigenvalue crossings excluded), and matches the published eigenframe
# dynamics (Meneveau Annu. Rev. Fluid Mech. 2011 form: dλ_i/dt = −λ_i² + ¼(|ω|²−ω̃_i²) − P_ii^eig + visc).
#
# Evidence class: algebraic (exact symbolic). Scope: a kinematic/structural identity of the
# NS equations — NOT a regularity statement. :proved=0; distance UNTOUCHED. -- STABLE

import sympy as sp

ok_all = True
def report(name, expr_zero, detail=""):
    global ok_all
    ok = all(sp.simplify(e) == 0 for e in expr_zero)
    ok_all &= ok
    print(f"  [{'PASS' if ok else 'FAIL'}] {name}{('  — ' + detail) if detail else ''}")
    return ok

print("=" * 74)
print("  NS-046 sign Required-Check — Dλ₃ ⊃ −e₃ᵀ(∇²p)e₃  (exact, sympy)")
print("=" * 74)

# ── I1: gradient of NS ⇒ DA/Dt = −A² − ∇²p + νΔA ────────────────────────────
x, y, z, t, nu = sp.symbols('x y z t nu', real=True)
X = (x, y, z)
u = [sp.Function(f'u{i}')(x, y, z, t) for i in range(3)]
p = sp.Function('p')(x, y, z, t)

NS = [sp.diff(u[i], t) + sum(u[j] * sp.diff(u[i], X[j]) for j in range(3))
      + sp.diff(p, X[i]) - nu * sum(sp.diff(u[i], X[j], 2) for j in range(3))
      for i in range(3)]                                   # NS_i = 0 on solutions

A = sp.Matrix(3, 3, lambda i, k: sp.diff(u[i], X[k]))      # A_ik = ∂u_i/∂x_k
P = sp.Matrix(3, 3, lambda i, k: sp.diff(p, X[i], X[k]))   # pressure Hessian
DADt = sp.Matrix(3, 3, lambda i, k: sp.diff(A[i, k], t)
                 + sum(u[j] * sp.diff(A[i, k], X[j]) for j in range(3)))
LapA = sp.Matrix(3, 3, lambda i, k: sum(sp.diff(A[i, k], X[j], 2) for j in range(3)))
gradNS = sp.Matrix(3, 3, lambda i, k: sp.diff(NS[i], X[k]))

resid_I1 = (DADt + A * A + P - nu * LapA) - gradNS         # must vanish IDENTICALLY (no eqns assumed)
report("I1  ∂_k(NS_i) ≡ [DA/Dt + A² + ∇²p − νΔA]_ik", list(resid_I1),
       "so on NS solutions DA/Dt = −A² − ∇²p + νΔA")

# ── I2: sym(A²) = S² + Ω² for generic A ──────────────────────────────────────
a = sp.Matrix(3, 3, lambda i, j: sp.Symbol(f'a{i}{j}', real=True))
S = (a + a.T) / 2
Om = (a - a.T) / 2
resid_I2 = ((a * a + (a * a).T) / 2) - (S * S + Om * Om)
report("I2  sym(A²) = S² + Ω²", list(resid_I2),
       "⇒ DS/Dt = −S² − Ω² − ∇²p + νΔS (Hessian is symmetric)")

# ── I3: Ω² = ¼(ω⊗ω − |ω|²I), with ω built FROM the same generic A ────────────
w = sp.Matrix([a[2, 1] - a[1, 2], a[0, 2] - a[2, 0], a[1, 0] - a[0, 1]])  # ω_i = ε_ijk A_kj
I3_rhs = (w * w.T - (w.T * w)[0] * sp.eye(3)) / 4
resid_I3 = Om * Om - I3_rhs
report("I3  Ω² = ¼(ω⊗ω − |ω|²I)", list(resid_I3),
       "⇒ e₃ᵀ(−Ω²)e₃ = ¼(|ω|² − (ω·e₃)²) ≥ 0 (the vorticity FEED)")

# ── I4: eigen-derivative lemma dλ₃/dt = e₃ᵀ(dS/dt)e₃ on a parametrized family ─
# S(t) = R(t) diag(λ_i(t)) R(t)ᵀ with R(t) a one-parameter rotation composition (generic axis).
th1, th2 = sp.Function('theta1')(t), sp.Function('theta2')(t)
l1, l2, l3 = sp.Function('lam1')(t), sp.Function('lam2')(t), sp.Function('lam3')(t)
Rz = sp.Matrix([[sp.cos(th1), -sp.sin(th1), 0], [sp.sin(th1), sp.cos(th1), 0], [0, 0, 1]])
Rx = sp.Matrix([[1, 0, 0], [0, sp.cos(th2), -sp.sin(th2)], [0, sp.sin(th2), sp.cos(th2)]])
R = Rz * Rx
St = R * sp.diag(l1, l2, l3) * R.T
e3 = R[:, 2]                                               # unit eigenvector of λ₃ (R orthogonal)
lhs_I4 = sp.diff(l3, t)
rhs_I4 = (e3.T * sp.diff(St, t) * e3)[0]
report("I4  dλ₃/dt = e₃ᵀ(dS/dt)e₃ (λ₃ simple)", [lhs_I4 - rhs_I4],
       "Hellmann–Feynman; rotation terms cancel exactly")

# ── Assembly: project DS/Dt onto e₃ ──────────────────────────────────────────
# By I1+I2(+symmetry of P): DS/Dt = −S² − Ω² − P + νΔS. By I4: Dλ₃/Dt = e₃ᵀ(DS/Dt)e₃.
# A1: the self-term projection e₃ᵀS²e₃ = λ₃² — verified symbolically on the I4 family,
# where Se₃=λ₃e₃ holds by construction (the general case is the 1-line operator identity
# e₃ᵀS²e₃ = (Se₃)·(Se₃) = λ₃², S symmetric, |e₃|=1):
resid_A1 = (e3.T * St * St * e3)[0] - l3 ** 2
report("A1  e₃ᵀS²e₃ = λ₃²  (so e₃ᵀ(−S²)e₃ = −λ₃²)", [resid_A1],
       "the λ₃ 'self' term — note the sign: self-DAMPING for the max eigenvalue")

print("-" * 74)
if ok_all:
    print("  ⇒ Dλ₃/Dt = −λ₃² + ¼(|ω|² − (ω·e₃)²) − e₃ᵀ(∇²p)e₃ + ν e₃ᵀ(ΔS)e₃")
    print("  ⇒ the pressure-Hessian projection enters with coefficient −1:")
    print("    e₃ᵀ(∇²p)e₃ > 0  ⇒  DEPLETES the max-stretch growth.   CONVENTION PINNED.")
    print("  NOTE (honest sharpening): for λ₃ itself the −λ₃² term is SELF-DAMPING; the")
    print("  growth feed is the vorticity term ¼(|ω|²−(ω·e₃)²) (maximal for ω ⊥ e₃). The")
    print("  probes' ratio R = e₃ᵀ∇²p e₃ / λ₃² remains a sensible magnitude comparison")
    print("  (λ₃² is the dynamics' scale), but 'λ₃² = self-amplification OF λ₃' is loose.")
    print("  Caveat: I4 needs λ₃ simple (a.e.; crossings excluded). :proved=0.")
else:
    print("  ✗ A FAIL above — the convention is NOT pinned; re-read the probes' sign before use.")
print("=" * 74)
raise SystemExit(0 if ok_all else 1)

File: scripts/ns046_dlambda3_sign_check.jl
#!/usr/bin/env julia
# ns046_dlambda3_sign_check.jl — independent corroboration of the sympy sign Required-Check
# (scripts/ns046_dlambda3_sign_check.py). Two rungs:
#   (1) EXACT Rational{BigInt}: the matrix identities I2 (sym(A²)=S²+Ω²) and I3
#       (Ω² = ¼(ω⊗ω − |ω|²I)) on random exact-rational matrices — zero residual, not 1e-16.
#   (2) Float64 finite-difference CONVERGENCE for I4 + the assembled formula:
#       dλ₃/dt vs e₃ᵀ(dS/dt)e₃ on a smooth parametrized S(t) family — central-diff error
#       must shrink ~4× per h-halving (O(h²)) toward the formula value.
# Evidence class: algebraic (rung 1) + computational corroboration (rung 2). :proved=0. -- STABLE

using LinearAlgebra, Random, Printf

ok_all = true
function report(name, ok; detail="")
    global ok_all &= ok
    @printf("  [%s] %s%s\n", ok ? "PASS" : "FAIL", name, isempty(detail) ? "" : "  — "*detail)
end

println("="^74)
println("  NS-046 sign check — Julia corroboration (exact-rational + FD convergence)")
println("="^74)

# ── rung 1: EXACT identities on random Rational{BigInt} matrices ─────────────
rng = MersenneTwister(46)
rq() = Rational{BigInt}(rand(rng, -9:9), rand(rng, 1:9))
maxres = (M) -> maximum(abs.(M))
res2 = zero(Rational{BigInt}); res3 = zero(Rational{BigInt})
for trial in 1:25
    A = [rq() for _ in 1:3, _ in 1:3]
    S = (A + A') .// 2;  Om = (A - A') .// 2
    global res2 = max(res2, maxres((A*A + (A*A)') .// 2 .- (S*S .+ Om*Om)))
    ω = [A[3,2]-A[2,3], A[1,3]-A[3,1], A[2,1]-A[1,2]]
    I3rhs = (ω*ω' .- (ω'ω) .* Matrix{Rational{BigInt}}(I,3,3)) .// 4
    global res3 = max(res3, maxres(Om*Om .- I3rhs))
end
report("I2 exact (25 random Rational{BigInt} A): sym(A²)=S²+Ω²", res2 == 0; detail="max residual = $(res2) (exact zero)")
report("I3 exact: Ω² = ¼(ω⊗ω − |ω|²I)",                       res3 == 0; detail="max residual = $(res3) (exact zero)")

# ── rung 2: FD convergence for I4 + assembled Dλ₃ formula ────────────────────
# S(t) = R(t) diag(λ(t)) R(t)ᵀ, generic smooth λ_i(t) (kept separated ⇒ λ₃ simple), R(t) two-angle rotation.
λs(t) = (-1.0 - 0.3sin(t), 0.2 + 0.1cos(2t), 0.8 + 0.2sin(1.3t))   # λ₁<λ₂<λ₃ on the window
function Rt(t)
    θ1 = 0.7t + 0.2; θ2 = 0.4sin(t) + 0.1
    Rz = [cos(θ1) -sin(θ1) 0; sin(θ1) cos(θ1) 0; 0 0 1]
    Rx = [1 0 0; 0 cos(θ2) -sin(θ2); 0 sin(θ2) cos(θ2)]
    Rz*Rx
end
St(t) = (R=Rt(t); R*Diagonal(collect(λs(t)))*R')
λ3(t) = λs(t)[3]
e3t(t) = Rt(t)[:,3]
t0 = 0.9
dSdt_fd(h) = (St(t0+h) .- St(t0-h)) ./ (2h)
form(h)   = dot(e3t(t0), dSdt_fd(h)*e3t(t0))            # e₃ᵀ(dS/dt)e₃ (FD in S only)
dλ3_fd(h) = (λ3(t0+h) - λ3(t0-h)) / (2h)
errs = [abs(dλ3_fd(h) - form(h)) for h in (1e-2, 5e-3, 2.5e-3)]
rates = [errs[i]/errs[i+1] for i in 1:2]
conv_ok = all(r -> 3.0 < r < 5.0, rates) || maximum(errs) < 1e-10
report("I4 FD: dλ₃/dt = e₃ᵀ(dS/dt)e₃ converges O(h²)", conv_ok;
       detail=@sprintf("errs %.2e → %.2e → %.2e (rates %.2f, %.2f)", errs..., rates...))

# assembled formula on a synthetic 'NS-like' decomposition: given arbitrary symmetric P̃, Ω̃(ω), check
# the e₃-projection algebra  e₃ᵀ(−S²−Ω²−P̃)e₃ = −λ₃² + ¼(|ω|²−(ω·e₃)²) − e₃ᵀP̃e₃  (Float64, machine-ε)
S0 = St(t0); e3 = e3t(t0); l3 = λ3(t0)
ω0 = [0.37, -1.21, 0.64]
Om0 = [0 -ω0[3] ω0[2]; ω0[3] 0 -ω0[1]; -ω0[2] ω0[1] 0] ./ 2  # Ω_ij = −½ε_ijk ω_k
P0 = [1.1 0.2 -0.4; 0.2 -0.7 0.3; -0.4 0.3 0.9]
lhs = dot(e3, (-S0*S0 .- Om0*Om0 .- P0)*e3)
rhs = -l3^2 + 0.25*(dot(ω0,ω0) - dot(ω0,e3)^2) - dot(e3, P0*e3)
report("Assembly: e₃ᵀ(−S²−Ω²−P)e₃ = −λ₃²+¼(|ω|²−(ω·e₃)²)−e₃ᵀPe₃", abs(lhs-rhs) < 1e-12;
       detail=@sprintf("|lhs−rhs| = %.2e", abs(lhs-rhs)))

println("-"^74)
println(ok_all ?
    "  ⇒ CORROBORATED: the pressure-Hessian enters Dλ₃ with coefficient −1 (convention PINNED)." :
    "  ✗ FAIL above — do NOT treat the convention as pinned.")
println("="^74)
exit(ok_all ? 0 : 1)
